app/uk/gov/hmrc/gform/upscan/UpscanController.scala

```diff
@@ -0,0 +1,203 @@
+import uk.gov.hmrc.gform.sharedmodel.form.{ EnvelopeId, FileId, Form, FormIdData }
+import uk.gov.hmrc.gform.sharedmodel.formtemplate.{ FormComponentId, FormTemplateId, SectionNumber }
+import uk.gov.hmrc.http.HeaderCarrier
+import uk.gov.hmrc.play.bootstrap.frontend.controller.FrontendController
+
+class UpscanController(
+  auth: AuthenticatedRequestActions,
+  nonAuth: NonAuthenticatedRequestActionsAlgebra[Future],
+  fastForwardService: FastForwardService,
+  fileUploadFrontendAlgebra: FileUploadFrontendAlgebra[Future],
+  gformBackEndAlgebra: GformBackEndAlgebra[Future],
+  upscanService: UpscanAlgebra[Future],
+  queryParameterCrypto: CryptoWithKeysFromConfig,
+  i18nSupport: I18nSupport,
+  messagesControllerComponents: MessagesControllerComponents
+)(implicit ec: ExecutionContext, s: Scheduler)
+    extends FrontendController(messagesControllerComponents) with Retrying {
+
+  import i18nSupport._
+
+  private val logger: Logger = LoggerFactory.getLogger(getClass)
+
+  private def setTransferred(
+    form: Form,
+    formComponentId: FormComponentId,
+    fileId: FileId
+  ): Form =
+    form.copy(
+      componentIdToFileId = form.componentIdToFileId + (formComponentId, fileId)
+    )
+
+  def callback(formComponentId: FormComponentId, envelopeId: EnvelopeId, formIdDataCrypted: Crypted): Action[JsValue] =
+    nonAuth.json { implicit request => l =>
+      implicit val hc = HeaderCarrier() // This is called by upscan service, this call has no session data
+
+      val upscanCallbackPayload: JsResult[UpscanCallback] = request.body.validate[UpscanCallback]
+
+      val decrypted: PlainText = queryParameterCrypto.decrypt(formIdDataCrypted)
+
+      val formIdDataJsResult: JsResult[FormIdData] = Json.parse(decrypted.value).validate[FormIdData]
+      val fileId = FileId(formComponentId.value)
+
+      val callbackResult: Either[Unit, Future[Result]] = for {
+        formIdData <-
+          formIdDataJsResult.asEither.leftMap(error =>
+            logger.error(s"Upscan callback - formIdData failed to parse: $formIdDataJsResult, reported error: $error")
+          )
+        upscanCallback <-
+          upscanCallbackPayload.asEither.leftMap(error =>
+            logger.error(s"Upscan callback payload failed to parse: ${request.body}, reported error: $error")
+          )
+
+      } yield upscanCallback match {
+        case upscanCallbackSuccess: UpscanCallback.Success =>
+          for {
+            file <- upscanService.download(upscanCallbackSuccess.downloadUrl)
+            _    <- fileUploadFrontendAlgebra.uploadFile(envelopeId, fileId, upscanCallbackSuccess.uploadDetails, file)
+            form <- gformBackEndAlgebra.getForm(formIdData)
+            formUpd = setTransferred(form, formComponentId, fileId)
+            _                  <- gformBackEndAlgebra.updateUserData(formUpd, formIdData.maybeAccessCode)
+            upscanConfirmation <- upscanService.confirm(upscanCallbackSuccess)
+          } yield NoContent
```

Comment:
After updating the user data in forms collection, if upscan confirm fails, we need to rollback the update isnt it?